### PR TITLE
Block txns

### PR DESCRIPTION
**Pull request by marioiordanov**
based on #103 
Most of the files that are changed are coming from the PR that this one is based on
## Usage related changes

starknet_getBlockWithTxHashes
starknet_getBlockWithTxs

## Development related changes

Added:

- version to each transaction struct
- max_fee, signature, nonce, chain_id to transactions that uses StarknetInRust implementation of the corresponding transaction type, these properties are private in the StarknetInRust implementation
- getter methods
- utils.rs for converting between types

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [ ] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet/src/blocks/mod.rs [hunk 1]
```diff
@@ -54,12 +59,35 @@ impl StarknetBlocks {
...
+    ) {
+        self.num_to_state_diff.insert(block_number, state_diff);
+    }
+
+    pub fn get_by_block_id(&self, block_id: BlockId) -> Option<&StarknetBlock> {
+        match block_id {
+            BlockId::Hash(hash) => self.get_by_hash(Felt::from(hash)),
+            BlockId::Number(block_number) => self.num_to_block.get(&BlockNumber(block_number)),
+            BlockId::Tag(BlockTag::Latest) => {
+                if let Some(hash) = self.last_block_hash {
```
Comment (ivpavici) [inline crates/starknet/src/blocks/mod.rs:76]: you have solved the problem I had in my PR here :D

Comment (marioiordanov) [reply crates/starknet/src/blocks/mod.rs:76]: Sorry for not letting you know about it, you had spend your time going through the PR

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 2]
```diff
@@ -2,33 +2,152 @@ use starknet_core::error::Error;
...
+
+        Ok(Block {
+            status: *block.status(),
+            header: BlockHeader::from(&block),
+            transactions: crate::api::models::transaction::Transactions::Hashes(
+                block
+                    .get_transactions()
+                    .iter()
+                    // We shouldnt get in the situation where tx hash is None
+                    .map(|tx| FeltHex(tx.get_hash().unwrap_or_default()))
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/endpoints.rs:45]: Is this related to the situation in Mikolaj's PR: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/75#discussion_r1274983426

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 3]
```diff
@@ -2,33 +2,152 @@ use starknet_core::error::Error;
...
+        let block =
+            self.api.starknet.read().await.get_block(block_id.into()).map_err(|err| match err {
+                Error::NoBlock => ApiError::BlockNotFound,
+                unknown_error => ApiError::StarknetDevnetError(unknown_error),
+            })?;
+
+        let mut transactions = Vec::<TransactionWithType>::new();
+
+        for txn in block.get_transactions() {
+            let txn_to_add = match txn {
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/endpoints.rs:62]: Couldn't this conversion be extracted to a separate function? Or do we not want that?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet/src/transactions/invoke_transaction.rs [hunk 4]
```diff
@@ -35,22 +43,38 @@ impl InvokeTransactionV1 {
...
+                Felt::from(1).into(),
+                calldata.iter().map(|f| f.into()).collect(),
+                signature.iter().map(|f| f.into()).collect(),
+                chain_id.into(),
+                Some(nonce.into()),
+            )?,
+            chain_id,
+            signature,
+            nonce,
+            calldata,
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/transactions/invoke_transaction.rs:60]: they are present, but private

File: crates/starknet/src/transactions/mod.rs [hunk 5]
```diff
@@ -80,12 +80,57 @@ pub enum Transaction {
...
             Transaction::Declare(tx) => tx.transaction_hash,
             Transaction::DeclareV2(tx) => tx.transaction_hash,
-            Transaction::DeployAccount(tx) => Some(tx.0.hash_value().clone().into()),
-            Transaction::Invoke(tx) => Some(tx.0.hash_value().clone().into()),
+            Transaction::DeployAccount(tx) => Some(tx.inner.hash_value().clone().into()),
+            Transaction::Invoke(tx) => Some(tx.inner.hash_value().clone().into()),
+        }
+    }
+
+    pub fn chain_id(&self) -> &Felt {
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/transactions/mod.rs:92]: Yes, there are, your suggestion is one of them. But couldn't decide which one to use.

[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/transactions/mod.rs:92]: Ok, I guess we can then leave it as it is.

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:45]: I thought it was fixed in his PR, but here it is the fix [fa6dfd6](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/105/commits/fa6dfd647fd6642470003c2750fa1a7ab462b273)

Comment (FabijanC) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:45]: Well his PR hasn't been merged yet

Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:45]: this fix is not present in his PR

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 6]
```diff
@@ -2,33 +2,152 @@ use starknet_core::error::Error;
...
+                    .map(|tx| FeltHex(tx.get_hash().unwrap_or_default()))
+                    .collect(),
+            ),
+        })
     }
 
     /// starknet_getBlockWithTxs
-    pub(crate) async fn get_block_with_full_txs(&self, _block_id: BlockId) -> RpcResult<Block> {
-        Err(error::ApiError::BlockNotFound)
+    pub(crate) async fn get_block_with_full_txs(&self, block_id: BlockId) -> RpcResult<Block> {
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: addressed in [606c565](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/105/commits/606c565cfedf26bcc11f0bf046cabf1f649aa72b)

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 3 above)
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: Addressed in [3eac7ef](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/105/commits/3eac7ef888625caf0317141dd258aaa8cbcf11e7)

**Review by FabijanC — APPROVED** [review]
Approving, can be squashed & merged after resolving conflicts.

I linked issue #95 as closeable by this PR since PR #113 was not merged into main.

**Review by ivpavici — APPROVED** [review]

**Pull request merged**